http/relay/relay_test.go

```diff
@@ -87,18 +133,100 @@ func TestRelay(t *testing.T) {
 		})
 	})
 
+	closeApp := func(t *testing.T) {
+		t.Helper()
+		r1 := numRestarts()
+		body := newRequestBody(t, "GET", "/close", nil)
+		assertString(t, "closed", body)
+
+		r2 := numRestarts()
+		assert.Equal(t, r2, r1+1)
+	}
+
+	// A bad route (such as /throw) eventually stops retrying
+	t.Run("bad route", func(t *testing.T) {
+		newH(t)
+
+		r1 := numRestarts()
+		res := httptest.NewRecorder()
+		req := httptest.NewRequest("GET", "/throw", nil)
+		h.ServeHTTP(res, req)
+		r2 := numRestarts()
+
+		// Be a good proxy now
+		assert.Equal(t, 502, res.Code)
+
+		// 3 retries == 4 tries
+		assert.Equal(t, r2-r1, 4)
+	})
+
 	t.Run("restart server", func(t *testing.T) {
+		newH(t)
+
 		t.Run("200", func(t *testing.T) {
-			res := httptest.NewRecorder()
-			req := httptest.NewRequest("GET", "/throw/env", nil)
+			body := newRequestBody(t, "GET", "/throw/env", nil)
+			assertString(t, "Hello", body)
+		})
 
-			h.ServeHTTP(res, req)
+		t.Run("use after close", func(t *testing.T) {
+			closeApp(t)
+		})
 
-			assert.Equal(t, 200, res.Code)
-			assertString(t, "Hello", res.Body.String())
+		t.Run("use after close restart loop", func(t *testing.T) {
+			for i := 0; i < 7; i++ {
+				closeApp(t)
+			}
 		})
 	})
 
+	t.Run("child process cleanup", func(t *testing.T) {
+
+		// In this test, we test that a child process who stops accepting network connections
+		// (but hasn't crashed) is gracefully asked to be shut down
+		t.Run("net offline zombie", func(t *testing.T) {
+			newH(t)
+
+			pid := childPid(t)
+			process, err := os.FindProcess(pid)
+			assert.NoError(t, err, "find process")
+
+			start := time.Now()
+			closeApp(t)
+			ps, err := process.Wait()
+			if err != nil {
+				// This process might have completed before this test progressed this far
+				assert.Contains(t, err.Error(), "no child processes")
+				return
+			}
+
+			assert.True(t, time.Since(start).Seconds() >= 10)
+			assert.True(t, time.Since(start).Seconds() < 12)
+			assert.NoError(t, err, "zombie wait")
+			assert.True(t, ps.Exited())
+		})
+
+		// In this test, we test that a child process who swallows the "nice" shutdown signal
+		// will eventually be sent a SIGKILL and shut down
```

Comment:
I think this is a nice thing up can do, if your clean shutdown isn't working.  I think all Lambda processes need to be [kill-9](https://www.youtube.com/watch?v=Fow7iUaKrq4) safe anyway